src/cacao.ts

```diff
@@ -0,0 +1,111 @@
+import * as dagCbor from '@ipld/dag-cbor'
+import * as multiformats from 'multiformats'
+import * as multihashes from 'multihashes'
+import * as sha256 from '@stablelib/sha256'
+import { SiweMessage } from './siwe'
+
+const DAG_CBOR_CODEC_CODE = 113
+
+export type Header = {
+  t: string
+}
+
+export type Payload = {
+  domain: string
+  iss: string
+  chainId: number
+  aud: string
+  version: string
+  nonce: string
+  iat: number
+  nbf?: number
+  exp?: number
+  statement?: string
+  requestId?: string
+  resources?: Array<string>
+}
+
+export type Signature = {
+  s: string
+}
+
+export type Cacao = {
+  h: Header
+  p: Payload
+  s: Signature
+}
+
+export interface CacaoBlock {
+  cacao: Cacao
+  cid: multiformats.CID
+  bytes: Uint8Array
+}
+
+class CACAO {
+  cacao: Cacao
+  constructor(siwe: string | SiweMessage) {
+    const siweMessage = typeof siwe === 'string' ? new SiweMessage(siwe) : siwe
+    if (!siweMessage.signature) {
+      throw new Error('SIWE Message has not been signed yet')
+    }
+    this.cacao = this.fromSiweMessage(siweMessage)
+  }
+
+  async toCacaoBlock(): Promise<CacaoBlock> {
```

Comment:
I think this can be implemented in a simpler way just using the Block api: https://github.com/multiformats/js-multiformats/#creating-blocks

Reply:
o god - spent so much time in the dag-cbor docs, didnt think this would be in the multiformats package.

Reply:
Well, now you know how the byte layout looks like and how CIDs work :P

Reply:
welp, this simplified the API a lot.

Reply:
https://github.com/ceramicnetwork/cacao/pull/1/commits/16106bd4a1cb0cb10bae46c46cc508ca29609fcf

Reply:
nice